Add support for aggregations over distinct inputs to StreamingAggregation

duanmeng opened the pull request:
Fixes #7666

mbasmanova commented on velox/exec/StreamingAggregation.cpp:
```diff
@@ -104,11 +115,15 @@ bool equalKeys(
-    rows_->initializeRow(group, true);
-    storeKeys(group, index);
-    return group;
+  // TODO: Fix the issue that RowContainer::initializeRow would failed when
```
Is there a GitHub issue that describes this problem? Should we fix it first before introducing this PR?

duanmeng replied:
@mbasmanova Hi Masha, thanks for your review and guidance.
>Is there a GitHub issue that describes this problem?

No, this only happens when enabling disaggregation and the **reuse flag** is true in this PR that the `accumulator.destroy(rows)` in `RowContainer::freeAggregates` would throw an error,
```C++
E20240103 08:39:44.107774 3532646 Exceptions.h:68] Line: /Users/macduan/work/native/velox-8/velox/common/memory/HashStringAllocator.cpp:426, Function:free, Expression: !header->isFree() , Source: RUNTIME, ErrorCode: INVALID_STATE
unknown file: Failure
C++ exception with description "Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Retriable: False
Expression: !header->isFree()
Function: free
File: /Users/macduan/work/native/velox-8/velox/common/memory/HashStringAllocator.cpp
```

>Should we fix it first before introducing this PR?

OK, I will keep investigating this issue first.

duanmeng replied:
@mbasmanova Hi Masha, I found the root cause of this failure. I've updated the PR to fix it. Could you please help take a look? Thanks.

- The `DistinctAggregations::extractValues` calls the `aggregate.function->destroy(groups)` to destroy the groups at the end, which would call **`ArrayAggregate::desctroy`**.
- The `StreamingAggregation::startNewGroup` calls `RowContainer::initializeRow`, which calls `RowContainer::freeAggregates`. `RowContainer::freeAggregates` loops the `RowContainer::accumulators_`, and calls `accumulator.destroy(rows)` , which would call **`ArrayAggregate::desctroy`** too.
- The same memory would be freed twice, and cause the error.

```C++
// DistinctAggregations::extractValues
void extractValues(folly::Range<char**> groups, const RowVectorPtr& result)
      override {
    ...
      // Release memory back to HashStringAllocator to allow next
      // aggregate to re-use it.
      aggregate.function->destroy(groups);
    }
  }

char* StreamingAggregation::startNewGroup(vector_size_t index) {
  if (numGroups_ < groups_.size()) {
    auto group = groups_[numGroups_++];
    rows_->initializeRow(group, true);

// RowContainer::initializeRow
char* RowContainer::initializeRow(char* row, bool reuse) {
  if (reuse) {
    auto rows = folly::Range<char**>(&row, 1);
    freeVariableWidthFields(rows);
    freeAggregates(rows);

void RowContainer::freeAggregates(folly::Range<char**> rows) {
  for (auto& accumulator : accumulators_) {
    accumulator.destroy(rows);
  }
}

// ArrayAggregate::desctroy
void destroy(folly::Range<char**> groups) override {
    for (auto group : groups) {
      value<ArrayAccumulator>(group)->elements.free(allocator_);
    }
  }

```

mbasmanova replied:
Here is how we deal with the same issue in SortedAggregations. Perhaps, same "fix" can be applied to DistinctAggregations::extractValues.

```
void SortedAggregations::extractValues(
    folly::Range<char**> groups,
    const RowVectorPtr& result) {
  raw_vector<int32_t> temp;
...
    for (const auto& aggregate : aggregates) {
      aggregate->function->extractValues(
          groups.data(), groups.size(), &result->childAt(aggregate->output));

      // Release memory back to HashStringAllocator to allow next aggregate to
      // re-use it.
      aggregate->function->destroy(groups);
      // Overwrite empty groups over the destructed groups to keep the container
      // in a well formed state.
      aggregate->function->initializeNewGroups(
          groups.data(),
          folly::Range<const int32_t*>(
              iota(groups.size(), temp), groups.size()));
    }
```

duanmeng replied:
It works, thanks for your help.

mbasmanova commented on velox/exec/StreamingAggregation.cpp:
```diff
@@ -318,27 +355,32 @@ void StreamingAggregation::initializeNewGroups(size_t numPrevGroups) {
     function->setAllocator(&rows_->stringAllocator());
 
-    const auto rowColumn = rows_->columnAt(numKeys + i);
+    const auto rowColumn = rows_->columnAt(numKeys++);
```
modifying input arguments is an anti pattern; let's not do that

duanmeng replied:
Fixed

mbasmanova reviewed (commented):
@duanmeng Thank you for adding this functionality. Have a question below and also wonder if it makes sense to extend Aggregation Fuzzer to cover this use case.

mbasmanova commented:
@duanmeng Would you add "Fixes #xxx" comment to PR description to link this PR to the GitHub issue and allow that issue to auto-close when PR is merged?

mbasmanova commented on velox/exec/tests/StreamingAggregationTest.cpp:
```diff
@@ -146,6 +146,43 @@ class StreamingAggregationTest : public OperatorTestBase {
 
+  void testDistinctAggregation(
+      const std::vector<VectorPtr>& keys,
+      uint32_t outputBatchSize = 1'024) {
```
perhaps, remove default value to make it easier to read the code (no need to guess what the value is)

mbasmanova commented on velox/exec/tests/StreamingAggregationTest.cpp:
```diff
@@ -385,22 +422,17 @@ TEST_F(StreamingAggregationTest, sortedAggregations) {
-      makeFlatVector<int64_t>(1'000, [](auto row) { return row / 4; }),
-      makeFlatVector<int64_t>(1'000, [](auto row) { return row % 3; }),
-  });
+  auto size = 128;
```
Why 123 and not say 1'000?

duanmeng replied:
128 & 1024 are the same scenario as they are all less or equal to the default output batch (1024, `testDistinctAggregation(keys)`) and greater than the smaller output batch (32, `testDistinctAggregation(keys, 32)`). So I used a smaller size the make the test run faster.

mbasmanova commented on velox/exec/tests/StreamingAggregationTest.cpp:
```diff
@@ -385,22 +422,17 @@ TEST_F(StreamingAggregationTest, sortedAggregations) {
-                      false)
-                  .planNode();
+  std::vector<VectorPtr> keys = {
+      makeFlatVector<int32_t>(size, [](auto row) { return row; }),
```
Are we testing with single-row groups only? Is this intentional?

duanmeng replied:
No, I just copied it from other tests. I will add multikey test cases.

mbasmanova reviewed (commented):
@duanmeng The code changes looks good to me, but have some questions about testing. Also, still "wonder if it makes sense to extend Aggregation Fuzzer to cover this use case."

duanmeng commented:
@mbasmanova Hi Masha, I've resolved the latest comments and added the streaming distinct fuzzer test along with its stats. Could you help to take another look? Thanks.

mbasmanova commented on velox/exec/fuzzer/AggregationFuzzer.cpp:
```diff
@@ -84,6 +84,9 @@ class AggregationFuzzer : public AggregationFuzzerBase {
     size_t numDistinctInputs{0};
 
+    // Number of iterations using streaming aggregations over distinct inputs.
+    size_t numStreamingDistinctInputs{0};
```
This new counter seems to introduce inconsistence. We could streaming aggs with distinct inputs, but not with inputs as is. I suggest to remove this counter for now or figure out how to add it in a consistent way.

duanmeng replied:
Got it, removed. I will figure it out and add the related stats separately.

duanmeng commented:
@mbasmanova Hi Masha, I've rebased and added a minor optimization by removing the unnecessary `function->extractValues` of the distinct aggregation in https://github.com/facebookincubator/velox/pull/8176/commits/e9234a735e6b13de8196c19aa693296033e6308b. Could you please help take a look? Thanks a lot.

mbasmanova commented on velox/exec/tests/StreamingAggregationTest.cpp:
```diff
@@ -224,6 +278,84 @@ class StreamingAggregationTest : public OperatorTestBase {
     EXPECT_EQ(NonPODInt64::constructed, NonPODInt64::destructed);
   }
+
+  void testMultiKeyDistinctAggregation(
```
This tests doesn't seem to use streaming aggregation. Is this intentional?

duanmeng replied:
No, Fixed. I should use `PlanBuilder::streamingAggregation` explicitly and use `AggregationNode::Step::kSingle`.

mbasmanova reviewed (commented):
@duanmeng Looks great % one question about the test.

mbasmanova reviewed (approved):
Thanks.
